src/coreclr/jit/objectalloc.cpp

```diff
@@ -249,9 +258,30 @@ void ObjectAllocator::MarkEscapingVarsAndBuildConnGraph()
 
     for (BasicBlock* const block : comp->Blocks())
     {
+        const bool         basicBlockHasBackwardJump = block->HasFlag(BBF_BACKWARD_JUMP);
+        BitSetShortLongRep lclsInPreds               = BitVecOps::UninitVal();
+
+        if (basicBlockHasBackwardJump)
+        {
+            lclsInPreds = BitVecOps::MakeEmpty(&m_bitVecTraits);
+            for (BasicBlock* const predBlock : block->PredBlocks())
+            {
+                for (Statement* const stmt : predBlock->Statements())
+                {
+                    GenTree* stmtExpr = stmt->GetRootNode();
+                    if (stmtExpr->OperIsLocalStore())
+                    {
+                        GenTreeLclVarCommon* lclVar = stmtExpr->AsLclVarCommon();
+                        unsigned int         lclNum = lclVar->GetLclNum();
+                        BitVecOps::AddElemD(&m_bitVecTraits, lclsInPreds, lclNum);
+                    }
+                }
+            }
+        }
```

Comment:
Why is checking the preds like this sufficient? What prevents the allocated object from being stored in some other block inside the loop that reaches the backedge?

I think doing this analysis is going to require building the proper loop structures and probably also some form of liveness.

Reply:
> What prevents the allocated object from being stored in some other block inside the loop that reaches the backedge?

My thought is that when we see a backward edge, we check the local where the object is being stored to see if the target is a local defined in its preds, otherwise mark it escaping. For blocks that have back edges to the current block, they will be checked later so that we don't miss them. 
IMO this should be a conservative approximation without accurate loop and liveness.

Reply:
Agree with Jakob here.  I think the analysis is more complicated than this, and we need proper loop structure.

If you have a proper loop where the entry block dominates the body, then locals defined in the entry block can still be live on the back edges if they are first used before being defined.

Building loops and running liveness seems to be a necessary prerequisite, and unfortunately, liveness will be impaired at this stage of the jit, because we haven't yet figured out which locals are truly exposed, so we may end up also needing points-to analysis to help refine what can be aliased (and that likely should dovetail with or supplant the existing escape analysis), and proper loopness might also be a challenge given that we might end up creating a new loop after this phase because of a recursive tail call.

Assuming we can indeed soundly prove that an object's references cannot outlive a loop iteration (accounting for any possible liveness from EH), we'd also need to ensure that the object fields are zeroed (as they would be for a heap allocation) and/or that they are non-gc fields that are definitely written before they are read.

None of these problems are impossible, but it's a fair amount of work.